Deprecate registerFeature when using the main source set. (#27251)
gradle/gradle: platforms/documentation/docs/src/docs/userguide/releases/upgrading/upgrading_version_8.adoc

=== Deprecations

[[register_feature_main_source_set]]
==== Deprecated calling `registerFeature` using the `main` source set

Calling `link:{javadocPath}/org/gradle/api/plugins/JavaPluginExtension.html#registerFeature-java.lang.String-org.gradle.api.Action-[registerFeature]` on the `link:{javadocPath}/org/gradle/api/plugins/JavaPluginExtension.html[java]` extension using the `main` source set is deprecated and will change behavior in Gradle 9.0.

Currently, features created while calling `link:{javadocPath}/org/gradle/api/plugins/FeatureSpec.html#usingSourceSet-org.gradle.api.tasks.SourceSet-[usingSourceSet]` with the `main` source set are initialized differently than features created while calling `usingSourceSet` with any other source set.
Previously, when using the `main` source set, new `implementation`, `compileOnly`, `runtimeOnly`, `api`, and `compileOnlyApi` configurations were created, and the compile and runtime classpaths of the `main` source set were configured to extend these configurations.

> Starting in Gradle 9.0, the `main` source set will not be treated any differently than any other source set.

Reviewer: ```suggestion
Starting in Gradle 9.0, the `main` source set will be treated like any other source set.
```

Reviewer: Did I reword this correctly @jvandort? The current sentence is a mouth-full.

> With the `java-library` plugin applied, or any other plugin that applies the `java` plugin, this will effectively cause calling `usingSourceSet` with the `main` source set to throw an exception, as the `java` plugin already configures a main feature.

Reviewer: ```suggestion
With the `java-library` plugin applied (or any other plugin that applies the `java` plugin), calling `usingSourceSet` with the `main` source set will throw an exception.
This is because the `java` plugin already configures a main feature.
```
